### PR TITLE
community: Skip Login to HuggubgFaceHub when token is not set

### LINKED ISSUE #20342: Handling huggingfacehub_api_token=None for HuggingFaceEndpoint
### Checked other resources

- [X] I added a very descriptive title to this issue.
- [X] I searched the LangChain documentation with the integrated search.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I am sure that this is a bug in LangChain rather than my code.
- [X] The bug is not resolved by updating to the latest stable version of LangChain (or the specific integration package).

### Example Code

```
from langchain_community.llms import HuggingFaceEndpoint

llm = HuggingFaceEndpoint(
    endpoint_url="http://localhost:8010/",
    max_new_tokens=512,
    top_k=10,
    top_p=0.95,
    typical_p=0.95,
    temperature=0.01,
    repetition_penalty=1.03,
    huggingfacehub_api_token=None
)
```

### Error Message and Stack Trace (if applicable)

```
---------------------------------------------------------------------------
ValidationError                           Traceback (most recent call last)
Cell In[2], line 3
      1 from langchain_community.llms import HuggingFaceEndpoint
----> 3 llm = HuggingFaceEndpoint(
      4     endpoint_url="http://localhost:8010/",
      5     max_new_tokens=512,
      6     top_k=10,
      7     top_p=0.95,
      8     typical_p=0.95,
      9     temperature=0.01,
     10     repetition_penalty=1.03,
     11     huggingfacehub_api_token=None
     12 )

File ~/Jupyter/llm/venv/lib/python3.10/site-packages/langchain_core/load/serializable.py:120, in Serializable.__init__(self, **kwargs)
    119 def __init__(self, **kwargs: Any) -> None:
--> 120     super().__init__(**kwargs)
    121     self._lc_kwargs = kwargs

File ~/Jupyter/llm/venv/lib/python3.10/site-packages/pydantic/v1/main.py:341, in BaseModel.__init__(__pydantic_self__, **data)
    339 values, fields_set, validation_error = validate_model(__pydantic_self__.__class__, data)
    340 if validation_error:
--> 341     raise validation_error
    342 try:
    343     object_setattr(__pydantic_self__, '__dict__', values)

ValidationError: 1 validation error for HuggingFaceEndpoint
__root__
  Could not authenticate with huggingface_hub. Please check your API token. (type=value_error)
```

#### Background
While restructuring our codebase in response to the deprecation of `HuggingFaceTextGenInference`, I encountered an error when attempting to create a `HuggingFaceEndpoint` with a locally hosted [TGI server](https://github.com/huggingface/text-generation-inference). 

#### Issue
The error occurs in the `validate_environment function` of the `huggingface_endpoint.py` file, specifically in the lines [170-179](https://github.com/langchain-ai/langchain/blob/master/libs/community/langchain_community/llms/huggingface_endpoint.py#L170-L179).

The `@root_validator()` decorator throws an error when `huggingfacehub_api_token` is passed as `None`, which happens due to `login(token=huggingfacehub_api_token)` in `validate_environment` function. By commenting out the block that processes the API token and manually setting `huggingfacehub_api_token` to `None`, I am able to successfully create an `InferenceClient`.

Since HuggingFaceTextGenInference is fused into HuggingFaceEndpoint in PR #17254, we need to add logic to handle cases where `huggingfacehub_api_token` is passed as `None` or when no environment variable `HUGGINGFACEHUB_API_TOKEN` is set. This is particularly necessary for setups using a locally hosted TGI server where authentication with the Huggingface Hub may not be required.

### System Info

huggingface-hub==0.22.2
langchain-commnity==0.0.32

platform: linux
python version: 3.10

### PR DESCRIPTION
**Pull request by mirkenstein**
Thank you for contributing to LangChain!

- [ ] **HuggingFaceEndpoint**: "Skip Login to HuggingFaceHub"
  - Where:  langchain, community, llm, huggingface_endpoint
 

- [ ] **PR message**: ***Delete this entire checklist*** and replace with
    - **Description:** Skip login to huggingface hub when when `huggingfacehub_api_token` is not set. This is needed when using custom `endpoint_url` outside of HuggingFaceHub.
    - **Issue:** the issue # it fixes https://github.com/langchain-ai/langchain/issues/20342 and https://github.com/langchain-ai/langchain/issues/19685
    - **Dependencies:** None

- [ ] **Add tests and docs**: 
  1. Tested with locally available TGI endpoint
  2.  Example Usage
```python
from langchain_community.llms import HuggingFaceEndpoint

llm = HuggingFaceEndpoint(
    endpoint_url='http://localhost:8080',
    server_kwargs={
        "headers": {"Content-Type": "application/json"}
    }
)
resp = llm.invoke("Tell me a joke")
print(resp)
```
 Also tested against HF Endpoints
 ```python
 from langchain_community.llms import HuggingFaceEndpoint
huggingfacehub_api_token = "hf_xyz"
repo_id = "mistralai/Mistral-7B-Instruct-v0.2"
llm = HuggingFaceEndpoint(
    huggingfacehub_api_token=huggingfacehub_api_token,
    repo_id=repo_id,
)
resp = llm.invoke("Tell me a joke")
print(resp)
 ```
Additional guidelines:
- Make sure optional dependencies are imported within a function.
- Please do not add dependencies to pyproject.toml files (even optional ones) unless they are required for unit tests.
- Most PRs should not touch more than one package.
- Changes should be backwards compatible.
- If you are adding something to community, do not re-import it in langchain.

If no one reviews your PR within a few days, please @-mention one of baskaryan, efriis, eyurtsev, hwchase17.

### CODE REVIEW

**Comment by nwmoon** [comment]
it is necessary function !!

**Comment by mirkenstein** [comment]
Mentioning @mspronesti and @hwchase17 since you were the reviewers on my other HuggingFace PR https://github.com/langchain-ai/langchain/pull/21282

**Comment by miromihaylov** [comment]
@eyurtsev

**Comment by mirkenstein** [comment]
@eyurtsev

File: libs/community/langchain_community/llms/huggingface_endpoint.py
Comment (mspronesti) [inline libs/community/langchain_community/llms/huggingface_endpoint.py:185]: Hi @mirkenstein,
I'm not an official maintainer of langchain, but I will gladly give you my opinion -- how about we just allow the token to be `None` and attempt a login if passed? This would also require a try-catch block in `call` to rethrow the unauth error possibly thrown from huggingface hub if the token were to be needed.

Comment (mspronesti) [reply libs/community/langchain_community/llms/huggingface_endpoint.py:185]: Regarding the latter point, one should just replace the call to `get_from_dict_or_env` (which doesn't allow `None` values) with
```py
huggingfacehub_api_token = values["huggingfacehub_api_token"] or os.getenv("HUGGINGFACEHUB_API_TOKEN")
```
As for the first point, I need to double check but I believe `None` values are fine when instantiating hfhub's `InferneceClient`.

Comment (mirkenstein) [reply libs/community/langchain_community/llms/huggingface_endpoint.py]: It worked. See my latest commit.

**Comment by mirkenstein** [comment]
There is a new official library released from Huggning Face 
https://huggingface.co/blog/langchain
The new library still has the same issue.

Here is the deprecation warning when using the current community library.
```
LangChainDeprecationWarning: The class `HuggingFaceEndpoint` was deprecated in LangChain 0.0.37 and will be removed in 0.3. An updated version of the class exists in the from langchain-huggingface package and should be used instead. To use it run `pip install -U from langchain-huggingface` and import as `from from langchain_huggingface import llms import HuggingFaceEndpoint`.
  warn_deprecated(
 ```

File: libs/community/langchain_community/llms/huggingface_endpoint.py
Comment (mspronesti) [inline libs/community/langchain_community/llms/huggingface_endpoint.py:184]: These lines are leftovers, I suppose. Also, I guess a try-catch is still needed to re-throw the exception that the client might raise if the token was required, but not provided.

**Comment by mirkenstein** [comment]
> There is a new official library released from Huggning Face https://huggingface.co/blog/langchain The new library still has the same issue.
> 
> Here is the deprecation warning when using the current community library.
> 
> ```
> LangChainDeprecationWarning: The class `HuggingFaceEndpoint` was deprecated in LangChain 0.0.37 and will be removed in 0.3. An updated version of the class exists in the from langchain-huggingface package and should be used instead. To use it run `pip install -U from langchain-huggingface` and import as `from from langchain_huggingface import llms import HuggingFaceEndpoint`.
>   warn_deprecated(
> ```

The current library is not deprecated yet and it looks like it will be around for quite some time.  We can make this change also in the new `langchain-huggingface`  https://github.com/langchain-ai/langchain/blob/master/libs/partners/huggingface/langchain_huggingface/llms/huggingface_endpoint.py  as soon as this one gets merged.

**Review by morgandiverrez — APPROVED** [review]

**Comment by matthewgrossman** [comment]
@mirkenstein Any reason to not just bundle that change into this PR as well? Would help keep the change unified across both clients, easier to track down the original PR, etc

**Comment by mirkenstein** [comment]
> > @mirkenstein Any reason to not just bundle that change into this PR as well? Would help keep the change unified across both clients, easier to track down the original PR, etc
> 
> One of the guidelines states that "Most PRs should not touch more than one package" I applied the same change to the partner library and now the checks are failing at "CI / CI Success (pull_request)" step.

Created  https://github.com/langchain-ai/langchain/pull/22365 for the new partner client.

**Comment by Alex-J-W** [comment]
Will this also get merged into `libs/community/langchain_community/llms/huggingface_endpoint.py`? The partner package `langchain-huggingface` contains dependencies on torch and cuda packages that take too much disk space (for my needs).

**Comment by mirkenstein** [comment]
> Will this also get merged into `libs/community/langchain_community/llms/huggingface_endpoint.py`? The partner package `langchain-huggingface` contains dependencies on torch and cuda packages that take too much disk space (for my needs).

@efriis  and   @morgandiverrez

**Comment by matthewgrossman** [comment]
We switched internally to `langchain-huggingface` to pick up this auth-token-fix, and have noticed issues due to the increased package size as well. Because of this, we'd also prefer to use the `langchain-community` variant of the basellms that this PR fixes.

--- 

Alternatively, can the `langchain-huggingface` package be split up via some `extras`-ish functionality? The issue AFAICT is that `langchain-huggingface` has `embeddings/` which will import `sentence-transformers`:

```
 ~/test  » python3 -m venv venv; source venv/bin/activate
(venv)  ~/test  » pip install langchain_huggingface pipdeptree
... omitted for brevity ...
(venv)  ~/test  » pipdeptree --reverse --packages torch
torch==2.3.1
└── sentence-transformers==3.0.1 [requires: torch>=1.11.0]
    └── langchain-huggingface==0.0.3 [requires: sentence-transformers>=2.6.0]
```
My needs involve just getting a langchain-client to talk to TGI endpoints, and don't need embeddings at all.

I grepped over the many `pyproject.toml`s and didn't really find any examples of this (other than test dep groups, style, etc). If the langchain/huggingface maintainers are amenable to that change, can create a separate issue to track that work.

**Review by efriis — APPROVED** [review]

**Comment by efriis** [comment]
cc @Jofthomas on the ask to remove heavy dependencies for the just-cloud case!

**Review by efriis — APPROVED** [review]

**Pull request merged**
